service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/dataset/CreateTableCopyJobsStep.java

```diff
@@ -0,0 +1,194 @@
+    final FlightMap workingMap = flightContext.getWorkingMap();
+    // Gather inputs
+    final DatasetCloneInputs sourceInputs = getSourceInputs();
+    workingMap.put(ControlledResourceKeys.SOURCE_CLONE_INPUTS, sourceInputs);
+
+    final DatasetCloneInputs destinationInputs = getDestinationInputs(flightContext);
+    workingMap.put(ControlledResourceKeys.DESTINATION_CLONE_INPUTS, destinationInputs);
+
+    final AuthenticatedUserRequest userRequest =
+        flightContext
+            .getInputParameters()
+            .get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+    final BigQueryCow bigQueryCow = crlService.createBigQueryCow(userRequest);
+    // TODO:  remove usage of this client when it's all in CRL
+    final Bigquery bigQueryClient = crlService.createNakedBigQueryClient(userRequest);
+    try {
+      // Get a list of all tables in the source dataset
+      final TableList sourceTables =
+          bigQueryCow
+              .tables()
+              .list(sourceInputs.getProjectId(), sourceInputs.getDatasetName())
+              .execute();
+      // Start a copy job for each source table
+      final Map<String, String> tableToJobId =
+          Optional.ofNullable(
+                  workingMap.get(
+                      ControlledResourceKeys.TABLE_TO_JOB_ID_MAP,
+                      new TypeReference<Map<String, String>>() {}))
+              .orElseGet(HashMap::new);
+      for (TableList.Tables table : sourceTables.getTables()) {
+        if (tableToJobId.containsKey(table.getId())) {
+          // A job already exists for this table
+          continue;
+        }
+        // If the table contains data in its streaming buffer, that data can't be copied yet
+        // https://cloud.google.com/bigquery/streaming-data-into-bigquery#dataavailability
+        final Table tableGetResponse =
+            bigQueryCow
+                .tables()
+                .get(
+                    sourceInputs.getProjectId(),
+                    sourceInputs.getDatasetName(),
+                    getTableName(table.getId()))
+                .execute();
+        if (tableGetResponse.getStreamingBuffer() != null) {
+          // This is unfortunate, but our contract is to clone whatever BigQuery copies, which
+          // doesn't
+          // include recent streaming-inserted data. Warn in the log to assist in investigating
+          // customer complaints.
+          logger.warn(
+              "Streaming buffer data in table {} will not be copied. Estimated rows: {}, Oldest entry time: {}",
+              table.getId(),
+              tableGetResponse.getStreamingBuffer().getEstimatedRows(),
+              Instant.ofEpochMilli(
+                      tableGetResponse.getStreamingBuffer().getOldestEntryTime().longValueExact())
+                  .toString());
+        }
+        final Job inputJob = buildTableCopyJob(sourceInputs, destinationInputs, table);
+        // bill the job to the source project
```

Comment:
Should we bill the source or destination project? I'm thinking in terms of featured workspaces/resources, where (I think) it would make more sense for the destination to pay. That's probably a design-doc level discussion though

Reply:
Yes, we should make the destination pay.